### ISSUE TITLE
Suggestion: Run kubeadm from docker via a new `kubeadm --chroot=$rootfs`

### ISSUE THREAD

**Issue by anguslees** [kubernetes/kubeadm#503]
## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST

Currently most install instructions suggest to download the `kubeadm` tool onto the host filesystem either directly or via something like luxas' `kubeadm-installer` image (that just downloads a few binaries and copies them into the target filesystem).  This feels like an unnecessary step because I only need `kubeadm` briefly during node init/join, on my minimal filesystem I don't even _have_ wget/gunzip, and afterwards it's just wasted disk space.

Proposal: Add a `--chroot` flag to `kubeadm` and chroot to that directory early during `main()`.  Then we can publish `kubeadm` in a docker image and change the installation instructions to something like:

> Just run `docker run --rm -v /:/rootfs kubeadm:$version --chroot /rootfs`

Pros: fetching, versioning, updating, garbage collection are all handled for you via the regular docker machinery.

My assumption is that `kubeadm` wants to see the "real" filesystem for all its many file operations, and only wants to get the binary itself (and any dynamically linked libraries) from the docker image - hence the chroot suggestion.  An obvious alternative is to add an explicit directory prefix to all file operations in golang, rather than perform an actual chroot() - either is fine by me, it's just an intrusiveness vs surprise tradeoff.

**Comment by liyimeng** [comment]
+1

**Comment by luxas** [comment]
Why not do `chroot /rootfs -- kubeadm [args]`?

**Comment by anguslees** [comment]
> Why not do chroot /rootfs -- kubeadm [args]?

Because that will run `kubeadm` from `/rootfs`, which is not what we want.  We want to start the `kubeadm` binary from the docker image rootfs, and *then* chroot to the host rootfs.  The `chroot()` call must occur inside the `kubeadm` process for this to happen.

**Comment by kad** [comment]
@anguslees why do you want to start kubeadm against docker image rootfs ? Which steps you feel kubeadm should perform on image rootfs before starting to use real rootfs?

**Comment by anguslees** [comment]
@kad: We need to load the actual `kubeadm` executable (and any shared libraries if we should choose to link it dynamically) from the docker image rootfs.   Being able to distribute `kubeadm` as a docker image is the point of this proposal.

**Comment by kad** [comment]
@anguslees kubeadm is classical Go binary which means it has dependency only to libc.
Are you having something in your rootfs that doesn't meet that linking requirement ?

**Comment by anguslees** [comment]
> Are you having something in your rootfs that doesn't meet that linking requirement ?

No, this proposal is for the regular statically linked kubeadm binary.  I think I may have confused you with my description though, so let me try again:

We want the *kernel* to load the `kubeadm` executable from the image rootfs.  This means we can't chroot before the `exec()` call (this is what's wrong with luxas' suggestion above using the `chroot` tool).  We need the order to be: 1. `exec(kubeadm)`, 2. `chroot(/hostfs)`, 3. do application-level stuff with host file paths.

_After_ the kernel ELF loader has mmapped the `kubeadm` executable, etc, from the docker image _then_ this proposal is to `chroot()` to the host filesystem.  This results in a running `kubeadm` process that loaded the executable _from the docker image_ but can now manipulate files on the regular host filesystem as normal.  In particular, "as normal" means we don't need to use an explicit "/hostfs" path prefix everywhere (which is my alternate suggestion in the original comment).

Repeating the original motivation, the proposal is to have the `kubeadm` executable managed via docker (or other container runtime), because that provides a smoother/better/more-familiar install, upgrade, garbage-collect experience for the user than manually installing the `kubeadm` executable into the host filesystem itself.  (and it's ~5 lines of code)

**Comment by kad** [comment]
beside chroot() call, do you expect that kubeadm should handle mounting /proc and /dev/pts inside /rootfs or your container image will handle that ?

**Comment by anguslees** [comment]
@kad: I don't understand...  After the chroot() to the host `/`, `kubeadm` will be able to access the "real" `/proc` and `/dev/pts` as normal.  What am I missing?

**Comment by kad** [comment]
@anguslees sorry, my bad. You are correct. In case of passing real rootfs and chrooting to it, /proc, /dev will be visible. 

PID namespace will be different:
```shell
kad@kad:~> sleep 4000 &
[1] 5378
kad@kad:~> docker run -v /:/rootfs -it busybox chroot /rootfs kill 5378
kill: sending signal to 5378 failed: No such process
kad@kad:~>
```

Also kubeadm reset functionality would require privileged container to perform umounts, but tools like systemctl or docker executable should be working fine.

**Comment by anguslees** [comment]
@kad: `docker --pid=host ...`

**Comment by kad** [comment]
@anguslees so, let's rebase #54935 and get it in and mark as resolved ?

**Comment by anguslees** [comment]
(bug keepalive ping)
